USB: Fix reconnect issues

finikorg opened the pull request:
I still need to check suspend / resume properly working.

Fixes #15917

finikorg committed:
usb: device: Add transfer cancel helper

Add usb_cancel_transfers() helper to cancel all ongoing transfers.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: device: Do not call callback when transfer is cancelled

Only call callback when transfer is not cancelled.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: device: Cancel all transfers on disconnect

Cancel all usb transfers when receiving USB_DC_DISCONNECTED.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: device: Add debug logs catching re-enable failures

Add extra logs catching problem with re-enabling endpoints.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: webusb: Trivial syntax fix

Fix supended -> suspended.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: device: Disable endpoints on disconnect and suspend

Disable endpoints on following events: USB_DC_DISCONNECTED
and USB_DC_SUSPEND.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

rakons commented on subsys/usb/usb_device.c:
```diff
@@ -937,10 +943,51 @@ static void usb_register_status_callback(usb_dc_status_callback cb)
+{
+	u8_t ep = ep_data->ep_addr;
+
+	if (ep | !0x80) {
```
What it is supposed to do?

finikorg replied:
Sorry, my bad, removed that

rakons replied:
In general, it is good not to allow to disable EP0 and I suppose it was relating to that issue. I am not sure if the underlying library does take care of this?

finikorg replied:
That endpoint table contains only non control endpoints, so we do not need this (incorrect) check.

loicpoulain reviewed (commented):
LGTM

barsok reviewed (approved):
Found an issue which rakon then reported -- fixed. Verified on an own piece of HW.

masz-nordic reviewed (changes requested):
This allows USB3CV tests to be executed again, but following cases from Chapter 9 Tests still fail:

- TD 9.1, 9.2, 9.3, 9.4, 9.5, 9.7 in Configured state: 
  - `Set Configuration failed for Configuration Value : 1`
  - `(1.1.2) Devices must support being set to Addressed/Configured state.`

- TD 9.9, 9.13: 
  - `Set Configuration failed for Configuration Value : 1`
  - `(1.1.2) Devices must support being set to Addressed/Configured state.`

- TD 9.16:
  - `Enumeration failed at iteration 1 (General failure)`

finikorg commented:
> This allows USB3CV tests to be executed again, but following cases from Chapter 9 Tests still fail:
> 
>     * TD 9.1, 9.2, 9.3, 9.4, 9.5, 9.7 in Configured state:
>       
>       * `Set Configuration failed for Configuration Value : 1`
>       * `(1.1.2) Devices must support being set to Addressed/Configured state.`
> 
>     * TD 9.9, 9.13:
>       
>       * `Set Configuration failed for Configuration Value : 1`
>       * `(1.1.2) Devices must support being set to Addressed/Configured state.`
> 
>     * TD 9.16:
>       
>       * `Enumeration failed at iteration 1 (General failure)`

Are those failures resulting from the PR? Can they be reproduced on master?

masz-nordic commented:
On master the USB3CV tests do not work at all:

```
SetConfiguration() failed.
  No device:  deleting node!
  Device configuration failed!
  SetConfiguration() failed.
  No device:  deleting node!
  Device configuration failed!
A device was not found.  
Check that device is plugged into correct port.
Couldn't initialize common subsystems.
Function CVCommon_InitializeTestSuite did not succeed.  Return value = 0x80000000
```

finikorg commented:
> On master the USB3CV tests do not work at all:
> 
> ```
> SetConfiguration() failed.
>   No device:  deleting node!
>   Device configuration failed!
>   SetConfiguration() failed.
>   No device:  deleting node!
>   Device configuration failed!
> A device was not found.  
> Check that device is plugged into correct port.
> Couldn't initialize common subsystems.
> Function CVCommon_InitializeTestSuite did not succeed.  Return value = 0x80000000
> ```

OK, This is probably because of extra checks we have added. I think we should address this in a separate issue.

finikorg commented:
@masz-nordic Does this fix the issues?
```
diff --git a/subsys/usb/usb_device.c b/subsys/usb/usb_device.c
index 9e70966d10..765766cd5d 100644
--- a/subsys/usb/usb_device.c
+++ b/subsys/usb/usb_device.c
@@ -500,7 +500,6 @@ static bool set_endpoint(const struct usb_ep_descriptor *ep_desc)
 
        if (usb_dc_ep_enable(ep_cfg.ep_addr) < 0) {
                LOG_ERR("Failed to enable endpoint %x", ep_cfg.ep_addr);
-               return false;
        }
 
        usb_dev.configured = true;
```

masz-nordic commented:
@finikorg It fixes 9.1, 9.2, 9.3, 9.4, 9.5, 9.7 in Configured state and 9.9. But 9.13 and 9.16 still fail.
Additionally, there are new failures:
- 9.1, 9.2, 9.3, 9.4, 9.5, 9.7 in Default state
- 9.1, 9.2, 9.3, 9.4, 9.5, 9.7 in Addressed state

All of them fail with:
```
Set Configuration failed for Configuration Value : 0
(1.1.2) Devices must support being set to Addressed/Configured state.
```

finikorg commented:
Then we need to "fix" another bug :)
```
        if (usb_dc_ep_configure(&ep_cfg) < 0) {
                LOG_ERR("Failed to configure endpoint %x", ep_cfg.ep_addr);
-               return false;
        }
```

masz-nordic commented:
Still same failures as in previous case. Logs:
```
[00:01:11.285,858] <dbg> usb_hid.hid_custom_handle_req: Standard request: bRequest 0x9 bmRequestType 0x0 len 0
[00:01:11.285,858] <dbg> usb_device.usb_handle_std_device_req: REQ_SET_CONFIGURATION, conf 0x0
[00:01:11.285,858] <dbg> usb_device.usb_set_configuration: Device not configured - invalid configuration
[00:01:11.285,888] <dbg> usb_hid.hid_status_cb: cfg 0x200033a8 status 3
[00:01:11.285,888] <dbg> usb_hid.hid_do_status_cb: USB device configured
[00:01:11.285,888] <dbg> usb_device.usb_handle_std_device_req: USB Set Configuration failed
[00:01:11.285,888] <dbg> usb_device.usb_handle_request: Handler Error 0
[00:01:11.285,919] <dbg> usb_device.usb_print_setup: Setup: 0 9 0 0 0
[00:01:11.285,919] <dbg> usb_device.usb_handle_control_transfer: usb_handle_request failed
```

finikorg commented:
@masz-nordic what sample are you using? I had tests with frdm_k64f and HID it was passing all tests.

masz-nordic commented:
@finikorg `hid-mouse` with `nrf52840_pca10056`.

jfischer-phytec-iot commented:
Can you please apply and test again? (https://gist.github.com/acsplayon/b594d6b2207f8923edb071fdee499b36)

```diff
diff --git a/drivers/usb/device/usb_dc_nrfx.c b/drivers/usb/device/usb_dc_nrfx.c
index e3ba244b9c..86c7e39351 100644
--- a/drivers/usb/device/usb_dc_nrfx.c
+++ b/drivers/usb/device/usb_dc_nrfx.c
@@ -637,6 +637,8 @@ static void ep_ctx_reset(struct nrf_usbd_ep_ctx *ep_ctx)
        ep_ctx->buf.curr = ep_ctx->buf.data;
        ep_ctx->buf.len  = 0U;
 
+       ep_ctx->cfg.en = false;
+
        ep_ctx->read_complete = true;
        ep_ctx->read_pending = false;
        ep_ctx->write_in_progress = false;

```

masz-nordic commented:
@jfischer-phytec-iot This fixes 9.16. Other still fail, here is the trace from Default state tests:

![16193](https://user-images.githubusercontent.com/42532799/58013967-0bcd6900-7af8-11e9-95b0-064768529fcd.png)

If this is nRF specific issue and this PR fixes USB3CV for other platforms, let's merge it and I will investigate fails in a separate case.

finikorg commented:
I have tested frdm_k64f, stm32f723e_disco, nucleo_f412zg they are all passing, I will check also reel_board.

finikorg committed:
usb: device: Do not return error if ep already enabled

Fixes USB3CV Tool tests. One case is enabling endpoints with Set
Interface after Set Configuration. In this case we report warning and
continue without returning error.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
drivers: usb: usb_dc_nrfx: Set cfg.en to false on Reset

Set false to enable flag on USB Reset.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jfischer-no reviewed (approved):
LGTM

finikorg commented:
@masz-nordic Tests are  passing for `reel_board` also if I disable `USB_DEVICE_REMOTE_WAKEUP`

masz-nordic commented:
@finikorg: Which test mode are you using? I have tested `reel_board` with Compliance Test on Chapter 9 Tests [USB 2 devices] and it still fails.

finikorg commented:
> @finikorg: Which test mode are you using? I have tested `reel_board` with Compliance Test on Chapter 9 Tests [USB 2 devices] and it still fails.

Yes Compliance "Chapter 9 Tests [USB 2 devices]", only Remote Wakeup TD 9.12 is failing.

I am on top of my above mentioned PR, my reel_board is the oldest version, probably you have newer one?

jfischer-phytec-iot commented:
@masz-nordic Anything against to merge it?

masz-nordic commented:
@finikorg Mine is `1507.1`. I disabled remote wakeup in `Kconfig` (apart from `prj.conf`) too and now it only fails 9.21 and 9.16.

@jfischer-phytec-iot No, if this fixes problems from the original issue it is fine. I will create a separate one for USB3CV testcases.
